packages/utils/popup/usePopup.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  nextTick,
+  reactive,
+  toRefs,
+  onBeforeMount,
+  onBeforeUnmount,
+  onMounted,
+  getCurrentInstance,
+  watch,
+} from 'vue'
+import PopupManager from '../popup-manager'
+import getScrollBarWidth from '../scrollbar-width'
+import { getStyle, addClass, removeClass, hasClass } from '../dom'
+import isServer from '../isServer'
+
+let idSeed = 1
+
+let scrollBarWidth
+
+export const usePopup1 = {
+  watch: {
+    visible(val) {
+      if (val) {
+        if (this._opening) return
+        if (!this.rendered) {
+          this.rendered = true
+          nextTick(() => {
+            this.open()
+          }).then(() => {
+            //
+          })
+        } else {
+          this.open()
+        }
+      } else {
+        this.close()
+      }
+    },
+  },
+}
+
+const usePopup = (props, doClose) => {
+  let _popupId
+  let _opening = false
+  let _closing = false
+  let vm
+  let _closeTimer = null
+  let _openTimer = null
+  const state = reactive<State>({
+    opened: false,
+    bodyPaddingRight: null,
+    computedBodyPaddingRight: 0,
+    withoutHiddenClass: true,
+    rendered: false,
+    visible: false,
+  })
+
+  onMounted(() => {
+    vm = getCurrentInstance()
+  })
+
+  onBeforeMount(() => {
+    _popupId = 'popup-' + idSeed++
+    PopupManager.register(_popupId, vm)
+  })
+
+  onBeforeUnmount(() => {
+    PopupManager.deregister(_popupId)
+    PopupManager.closeModal(_popupId)
+    restoreBodyStyle()
+  })
+
+  const doOpen = merProps => {
+    if (isServer) return
+    if (vm.ctx.willOpen && !vm.ctx.willOpen()) return
```

Comment:
What's the purpose of `willOpen`? I didn't see any test for it.

Reply:
i removed it